pkg/storage/stores/shipper/bloomshipper/block_downloader.go

```diff
@@ -0,0 +1,233 @@
+	"github.com/grafana/dskit/services"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/loki/pkg/queue"
+	v1 "github.com/grafana/loki/pkg/storage/bloom/v1"
+	"github.com/grafana/loki/pkg/storage/stores/shipper/bloomshipper/config"
+	"github.com/grafana/loki/pkg/util"
+	"github.com/grafana/loki/pkg/util/constants"
+)
+
+type blockDownloader struct {
+	logger log.Logger
+
+	workingDirectory   string
+	queueMetrics       *queue.Metrics
+	queue              *queue.RequestQueue
+	blockClient        BlockClient
+	limits             Limits
+	activeUsersService *util.ActiveUsersCleanupService
+
+	ctx     context.Context
+	manager *services.Manager
+}
+
+func newBlockDownloader(config config.Config, blockClient BlockClient, limits Limits, logger log.Logger, reg prometheus.Registerer) (*blockDownloader, error) {
+	queueMetrics := queue.NewMetrics(reg, constants.Loki, "bloom_blocks_downloader")
+	//add cleanup service
+	downloadingQueue := queue.NewRequestQueue(config.BlocksDownloadingQueue.MaxTasksEnqueuedPerTenant, time.Minute, queueMetrics)
+	activeUsersService := util.NewActiveUsersCleanupWithDefaultValues(queueMetrics.Cleanup)
+
+	ctx := context.Background()
+	manager, err := services.NewManager(downloadingQueue, activeUsersService)
+	if err != nil {
+		return nil, fmt.Errorf("error creating service manager: %w", err)
+	}
+	err = services.StartManagerAndAwaitHealthy(ctx, manager)
+	if err != nil {
+		return nil, fmt.Errorf("error starting service manager: %w", err)
+	}
+
+	b := &blockDownloader{
+		ctx:                ctx,
+		logger:             logger,
+		workingDirectory:   config.WorkingDirectory,
+		queueMetrics:       queueMetrics,
+		queue:              downloadingQueue,
+		blockClient:        blockClient,
+		activeUsersService: activeUsersService,
+		limits:             limits,
+		manager:            manager,
+	}
+
+	for i := 0; i < config.BlocksDownloadingQueue.WorkersCount; i++ {
+		go b.serveDownloadingTasks(ctx, fmt.Sprintf("worker-%d", i))
+	}
+	return b, nil
+}
+
+type BlockDownloadingTask struct {
+	ctx   context.Context
+	block BlockRef
+	// ErrCh is a send-only channel to write an error to
+	ErrCh chan<- error
+	// ResultsCh is a send-only channel to return the block querier for the downloaded block
+	ResultsCh chan<- blockWithQuerier
+}
+
+func NewBlockDownloadingTask(ctx context.Context, block BlockRef, resCh chan<- blockWithQuerier, errCh chan<- error) *BlockDownloadingTask {
+	return &BlockDownloadingTask{
+		ctx:       ctx,
+		block:     block,
+		ErrCh:     errCh,
+		ResultsCh: resCh,
+	}
+}
+
+// noop implementation
+var onWorkerStopCallback = func() {}
+
+func (d *blockDownloader) serveDownloadingTasks(ctx context.Context, workerID string) {
+	logger := log.With(d.logger, "worker", workerID)
+	level.Debug(logger).Log("msg", "starting worker")
+
+	d.queue.RegisterConsumerConnection(workerID)
+	defer d.queue.UnregisterConsumerConnection(workerID)
+
+	idx := queue.StartIndexWithLocalQueue
+
+	for {
```

Comment:
it's not necessary here because the context will be checked during `Dequeue` function call...
also, we have discussed it here https://raintank-corp.slack.com/archives/C056KUQFBFT/p1700665187391789